File: solutions/typescript/028-diameter-of-binary-tree.ts
/**
 * Problem 28: Diameter of Binary Tree (LeetCode 543)
 * Difficulty: Easy
 * Language: TypeScript
 */
interface TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
}

function diameterOfBinaryTree(root: TreeNode | null): number {
  const dfs = (node: TreeNode | null): [number, number] => {
    // returns [height, diameter]
    if (!node) return [0, 0];
    const [lh, ld] = dfs(node.left);
    const [rh, rd] = dfs(node.right);
    return [1 + Math.max(lh, rh), Math.max(lh + rh, ld, rd)];
  };
  return dfs(root)[1];
}

File: solutions/ocaml/028-diameter-of-binary-tree.ml
(*
  Problem 28: Diameter of Binary Tree (LeetCode 543)
  Difficulty: Easy
  Language: OCaml
*)
type 'a tree = Nil | Node of 'a * 'a tree * 'a tree

let diameter_of_binary_tree root =
  let rec dfs = function
    | Nil -> (0, 0)
    | Node (_, l, r) ->
      let (lh, ld) = dfs l in
      let (rh, rd) = dfs r in
      (1 + max lh rh, List.fold_left max 0 [lh + rh; ld; rd])
  in
  snd (dfs root)
